libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/utils/ActiveLearningUtils.java

```diff
@@ -1,2 +1,104 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+
+package com.microsoft.bot.ai.qna.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.microsoft.bot.ai.qna.models.QueryResult;
+
+/**
+ * Active learning helper class.
+ */
+public class ActiveLearningUtils {
+
+    /**
+     * Previous Low Score Variation Multiplier.
+     */
+    private static final Float PREVIOUS_LOW_SCORE_VARIATION_MULTIPLIER = 0.7f;
+
+    /**
+     * Max Low Score Variation Multiplier.
+     */
+    private static final Float MAX_LOW_SCORE_VARIATION_MULTIPLIER = 1.0f;
+
+    private static Float MAXIMUM_SCORE_FOR_LOW_SCORE_VARIATION = 95.0f;
+
+    private static Float MINIMUM_SCORE_FOR_LOW_SCORE_VARIATION = 20.0f;
+
+    /**
+     * Gets maximum Score For Low Score Variation.
+     * @return Maximum Score For Low Score Variation.
+     */
+    public Float getMaximumScoreForLowScoreVariation() {
+        return ActiveLearningUtils.MAXIMUM_SCORE_FOR_LOW_SCORE_VARIATION;
+    }
+
+    /**
+     * Sets maximum Score For Low Score Variation.
+     * @param maximumScoreForLowScoreVariation Maximum Score For Low Score Variation.
+     */
+    public void setMaximumScoreForLowScoreVariation(Float maximumScoreForLowScoreVariation) {
+        ActiveLearningUtils.MAXIMUM_SCORE_FOR_LOW_SCORE_VARIATION = maximumScoreForLowScoreVariation;
+    }
+
+    /**
+     * Gets minimum Score For Low Score Variation.
+     * @return Minimum Score For Low Score Variation.
+     */
+    public Float getMinimumScoreForLowScoreVariation() {
+        return ActiveLearningUtils.MINIMUM_SCORE_FOR_LOW_SCORE_VARIATION;
+    }
+
+    /**
+     * Sets minimum Score For Low Score Variation.
+     * @param minimumScoreForLowScoreVariation Minimum Score For Low Score Variation.
+     */
+    public void setMinimumScoreForLowScoreVariation(Float minimumScoreForLowScoreVariation) {
+        ActiveLearningUtils.MINIMUM_SCORE_FOR_LOW_SCORE_VARIATION = minimumScoreForLowScoreVariation;
+    }
+
+    /**
+     * Returns list of qnaSearch results which have low score variation.
+     * @param qnaSearchResults List of QnaSearch results.
+     * @return List of filtered qnaSearch results.
+     */
+    public static List<QueryResult> getLowScoreVariation(List<QueryResult> qnaSearchResults) {
+        List<QueryResult> filteredQnaSearchResult = new ArrayList<QueryResult>();
+
+        if (qnaSearchResults.isEmpty()) {
```

Comment:
Shouldn't control if the list is null? I think if the list is null this part will not work. See [here](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs#L49)

Reply:
Agree with you